### DISCUSSION TITLE
Can't turn on/off the led light in program change?

### DISCUSSION THREAD

**Discussion by js-midi** [shanteacontrols/OpenDeck#102] [Q&A]
Can't turn on/off the led light in program change?
It turns on but doesn't turn off.
It works fine in control change, but in program change I can't turn the led on or off.
please tell me how

**Comment by paradajz** [comment]
Please try with the attached firmware - there was one issue with that. But normally, if the LED is configured to be controlled with Program change 0 for instance, you turn it off by switching to any other program change. Note: when you send program change to the board, the board switches preset first.

[blackpill401.zip](https://github.com/shanteacontrols/OpenDeck/files/7163062/blackpill401.zip)

**Reply by js-midi** [reply]
Still the button's led light is not updated with the newly pressed button's led.

**Reply by paradajz** [reply]
Please write me the entire flow. I'm not sure what you're doing.

**Reply by js-midi** [reply]
I am making a controller that changes the patch number of the sound module.
  It's a program changer

**Reply by paradajz** [reply]
Not that. The flow with the button presses and expectations regarding the output.

**Reply by js-midi** [reply]
I'm not very good at English, so it's difficult to explain in detail. I'm sorry.
We will place buttons to change the tone number of the sound module.
Also, when the button is pressed, the tone is selected and I want the button's led to light up.

**Reply by js-midi** [reply]
https://youtu.be/6vlB4NqrDuk

**Reply by js-midi** [reply]
The video content is using the cc button.
I want to implement the same with the program change button.

**Reply by paradajz** [reply]
Program change has only 2 values: channel and program value, unlike CC or Notes which have channel, ID and velocity or control value. Have you configured buttons so that each sends different program value?

**Reply by js-midi** [reply]
Yes !!

**Reply by paradajz** [reply]
Please send me your .sysex backup file.

**Reply by js-midi** [reply]
[OpenDeckUI-Backup-2021-09-16-13-04.zip](https://github.com/shanteacontrols/OpenDeck/files/7177911/OpenDeckUI-Backup-2021-09-16-13-04.zip)

Preset 1 is cc
Preset 2 is a program change

**Reply by paradajz** [reply]
Try with the attached firmware. It was a bug. This is still experimental, though, and I've also changed the touchscreen baudrate. It's now 38400. Please reflash the screen before loading this firmware. I've mentioned the new baudrate here: https://github.com/shanteacontrols/OpenDeck/wiki/Configuring-touchscreens.

[blackpill401.zip](https://github.com/shanteacontrols/OpenDeck/files/7180514/blackpill401.zip)

**Reply by js-midi** [reply]
After changing the baud rate, the screen response became slower.
The more pages are added, the slower it gets. From page 1 to page 8
The speed difference between page 1 and page 8 is huge.
page 8 is very slow

**Reply by paradajz** [reply]
That's interesting. I need to check that. What about program change?

**Reply by js-midi** [reply]
The program change is going very well
However, it is still not possible to set the initial screen value to a value other than 2 in the touch screen tab.
It only works well with the initial screen value of 2.

**Reply by paradajz** [reply]
I don't understand.

**Reply by paradajz** [reply]
By the way, I was testing my own controller with a touchscreen set to 38400 and didn't experience any slowdowns. Weird.

**Reply by js-midi** [reply]
maybe i'm doing something wrong
But there is no problem in using it.
After downgrading to 38400, it seems to be more stable.
thanks

**Reply by paradajz** [reply]
Can you send me a video showing the slow response?

**Reply by js-midi** [reply]
https://youtu.be/S56NsaE2i-0
https://youtu.be/JMYZyOFTlhg
When the first video is 115200
The second video is 38400

**Reply by paradajz** [reply]
Send me your configuration again please, with the current settings in slow version.

**Comment by js-midi** [comment]
[OpenDeckUI-Backup-2021-09-17-22-17.zip](https://github.com/shanteacontrols/OpenDeck/files/7188669/OpenDeckUI-Backup-2021-09-17-22-17.zip)

**Reply by paradajz** [reply]
Can I also check your touchscreen design? I haven't found anything so far...
What happens when you press the buttons on the touchscreen - are you sending them to PC? Is the board receiving any kind of messages from PC?

**Reply by js-midi** [reply]
[sd2prog.zip](https://github.com/shanteacontrols/OpenDeck/files/7188734/sd2prog.zip)

**Comment by paradajz** [comment]
I'm also sending you two firmware binaries from latest commit on master - just to confirm that this is (or isn't) baudrate issue. One is compiled for 38400 and the other for 115200. Once you load the one with 115200 you need to change it on the screen again.

[test.zip](https://github.com/shanteacontrols/OpenDeck/files/7188735/test.zip)

**Reply by js-midi** [reply]
yes

**Reply by paradajz** [reply]
I see in your design that you are using `page <n>` command to switch between pages. You can't use that since then firmware doesn't know which page is active. Also because of this, the icon changing cannot work due to this check in firmware:

```
 if ((_activeScreenID != icon.onScreen) && (_activeScreenID != icon.offScreen))
        return;    //don't allow setting icon on wrong screen
```

_activeScreenID is set to 1 the entire time since as far as the firmware is concerned, the page is never actually "switched" since there is no information on that. You need to assign printh 65... command for Touch Press event and then mark that button in the interface as page switching button (and also set the correct index of the page that is being switched to).

---

There is another, simpler alternative in that you can have only three pages in total:

0: logo
1: all icons in off state
2: all icons in on state

This requires you to redesign slightly.

Set the page 1 as the initial state, get rid of all "Button" objects and replace them all with "Hotspot" objects.

**Reply by paradajz** [reply]
Actually second option won't work since you need 64 buttons, ignore this then.

**Reply by js-midi** [reply]
Let's redesign
thank you ^^
